### PR TITLE
Light tiles now require apcs to provide light power

### PR DESCRIPTION
**Pull request by Qwertytoforty**
## What Does This PR Do

Light tiles turn off without apcs having lighting power enable, and can not be turned on without power.
Uses 100 W per tile. A lightbulb with light range of 4 uses 80 w, and this has a range of 5.
Light tiles must be manually turned on if turned off by power outage, do not self turn on, can be changed.

They also use the extinguish message when turned off via power as it is spookier.

## Why It's Good For The Game

Light tiles producing light in maints without requiring area power is not great for upcoming shadow demons. Or placed in space. Shadow demons can not tentacle the tiles either, meaning they can not deal with them.

This PR means that while still very effective vs shadow demons, they still require the apc to be powered. If the APC is depowered or destroyed, the demon will have much more freedom once again.

## Testing

Confirmed tiles worked.
Confirmed tiles turned off with no power
Confirmed tiles would not turn on with no power.
Confirmed message would not output each cycle.

## Changelog
:cl:
tweak: Light tiles now require apcs to provide light power in order to work.
/:cl:

### CODE REVIEW

File: code/game/turfs/simulated/floor/light_floor.dm [hunk 1]
```diff
@@ -25,8 +26,20 @@
 
 /turf/simulated/floor/light/BeforeChange()
 	set_light(0)
+	STOP_PROCESSING(SSobj, src)
 	..()
 
+/turf/simulated/floor/light/process()
+	if(power_check() || !on)
+		return
+	extinguish_light()
```
Comment (AffectedArc07) [inline code/game/turfs/simulated/floor/light_floor.dm:35]: We have a signal for area power changing or similar, look what light switches and stuff use.

*Please* dont make turfs process.

Comment (Qwertytoforty) [reply code/game/turfs/simulated/floor/light_floor.dm]: done

File: code/game/turfs/simulated/floor/light_floor.dm [hunk 2]
```diff
@@ -10,23 +10,51 @@
...
 	if(on)
 		icon_state = "light_on"
 		set_light(5, null, color)
+		if(!using_power)
+			A.addStaticPower(100, STATIC_LIGHT)
+			using_power = TRUE
 	else
 		icon_state = "light_off"
 		set_light(0)
+		if(using_power)
```
Comment (AffectedArc07) [inline code/game/turfs/simulated/floor/light_floor.dm:37]: Might wanna do this check on destroy as well

File: code/game/turfs/simulated/floor/light_floor.dm [hunk 3]
```diff
@@ -10,23 +10,51 @@
...
+			A.addStaticPower(-100, STATIC_LIGHT)
+			using_power = FALSE
+
```
Comment (hal9000PR) [inline code/game/turfs/simulated/floor/light_floor.dm:40]: not 100% happy with non-icon related logic being in an update_icon_state. I would appreciate it if you did a minor refactor and move this logic to `toggle_light`

File: code/game/turfs/simulated/floor/light_floor.dm (hunk 2 above)
Comment (Qwertytoforty) [reply code/game/turfs/simulated/floor/light_floor.dm:37]: Did some testing, as a turf is destroyed, before change is called, so the check is only needed on before change

Comment (AffectedArc07) [reply code/game/turfs/simulated/floor/light_floor.dm:37]: Destroying the turf will remove the area power drain?

File: code/game/turfs/simulated/floor/light_floor.dm (hunk 3 above)
Comment (Qwertytoforty) [reply code/game/turfs/simulated/floor/light_floor.dm]: done

File: code/game/turfs/simulated/floor/light_floor.dm (hunk 2 above)
Comment (Qwertytoforty) [reply code/game/turfs/simulated/floor/light_floor.dm]: Yes, as it calls BeforeChange, which turns off the light, removing the area power drain

File: code/game/turfs/simulated/floor/light_floor.dm [hunk 4]
```diff
@@ -71,11 +90,24 @@
 		to_chat(user, "<span class='warning'>[src]'s light bulb appears to have burned out.</span>")
 
 /turf/simulated/floor/light/proc/toggle_light(light)
+	if(!on && !power_check())
+		visible_message("<span class='danger'>[src] doesn't react, it seems to be out of power.</span>")
+		return
+	var/area/A = get_area(src)
+	if(!A)
+		return
```
Comment (hal9000PR) [inline code/game/turfs/simulated/floor/light_floor.dm:98]: this check isnt needed `power_check()` will fail if there is no area. (which it shouldnt do, if there is no area it should throw a runtime)

File: code/game/turfs/simulated/floor/light_floor.dm [hunk 5]
```diff
@@ -24,9 +31,21 @@
...
 /turf/simulated/floor/light/BeforeChange()
-	set_light(0)
+	toggle_light(FALSE)
 	..()
 
+/turf/simulated/floor/light/proc/power_update()
+	SIGNAL_HANDLER
+	if(power_check() || !on)
+		return
+	extinguish_light()
```
Comment (hal9000PR) [inline code/game/turfs/simulated/floor/light_floor.dm:41]: use `toggle_light`  extinguish light will spam messages if there is a bunch of tiles

File: code/game/turfs/simulated/floor/light_floor.dm [hunk 6]
```diff
@@ -24,9 +31,21 @@
...
+	SIGNAL_HANDLER
+	if(power_check() || !on)
+		return
+	toggle_light(FALSE)
+
+/turf/simulated/floor/light/proc/power_check()
+	var/area/A = get_area(src)
+	if(!A)
+		return FALSE
+	return A.powered(LIGHT)
```
Comment (hal9000PR) [inline code/game/turfs/simulated/floor/light_floor.dm:47]:
```suggestion
/turf/simulated/floor/light/proc/power_check()
	var/area/A = get_area(src)
	return A.powered(LIGHT)
```
this should also runtime

**Review by hal9000PR — APPROVED** [review]

**Pull request merged**